### DISCUSSION TITLE
Example for callbacks on DAG and task level

### DISCUSSION THREAD

**Discussion by p13rr0m** [apache/airflow#24894] [Q&A]
Hi everyone, 

Sorry for the inconvenience I have caused by creating the [issue](https://github.com/apache/airflow/issues/24868) before posting the question here. 

I would still like to follow up on it because I still don't fully understand the mechanism behind the DAG level on_failure_callback and on_success_callback.

Here is the documentation for both parameters of the DAG class. It states that this callback gets called when a DagRun fails and not when a TaskInstance fails.

```python
"""
:param on_failure_callback: A function to be called when a DagRun of this dag fails.
        A context dictionary is passed as a single parameter to this function.
:type on_failure_callback: callable
:param on_success_callback: Much like the ``on_failure_callback`` except
        that it is executed when the dag succeeds.
:type on_success_callback: callable
"""
```

When using this parameter, does this really mean all tasks from the DAG receive the callback? And is this the equivalent to setting the same parameters in the defaults_args parameter for the DAG object?

This discussion is probably related (https://github.com/apache/airflow/discussions/24029).

Thanks in advance for any help.

**Comment by potiuk** [comment]
You are probably referring to the https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/callbacks.html#example.

The example is correct. And my explanation was somewhat misleading.

Indeed "dag_failure_callback" is on DAG level and it will be fired once when it fails - with tha task that made it fail. so setting in on dag level is a shortcut. to set it to all tasks. In this simple example it is equivalent to setting the failure_callback on any task - because failure of any task will lead to the whole dag failure (remaining tasks will be skipped). Usually it is executed by the task that fails "technically". You could set "on_failure_callback" individually to all tasks in the DAG, but setting it in DAG level makes it simpler.

If you hava a more complex case, for example when you have a tasks that do not get skipped when a task gets fail using triggering rules, those two approaches are different. Setting on_task_failure  on individual task will fire when the task fails, setting on_task_failure on dag will fire when any of the "leaf" tasks in the DAG fail (and it will fire with the task that first failed there I believe - but looking at the code might give you better answers - I might be wrong when it comes to the sequence and there are likely edge-cases when there are multiple tasks failing the DAG potentially.

So the example is not wrong - it just uses the feature of DAG that any of the tasks failing, will trigger DAG failure.

**Comment by p13rr0m** [comment]
Thank you very much for your answer.

I see, so the statement "failures in any task call the task_failure_alert function" is actually a property of this specific example and not valid in a more general case. I think what makes this page so confusing to me is because in the section for [callback types](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/callbacks.html#callback-types) it mentions that the on_failure_callback is invoke when a tasks fails, which seems not to be the case when it is set on DAG level.

I have also checked the code to see how the callbacks on the DAG level are implemented, since I need to know if it gets called after a single leaf tasks fails or after all leaf task are done. It seems that it is the latter one. The update_state method in the DagRun class checks if all tasks are finished and then checks if there is any task that is in a failure state. If there is any, it calls the on_failure_callback directly.

**Reply by potiuk** [reply]
BTW. Since you made the investigation -  If you think you can improve the example and add more description - feel free to add clarification and updates to the page. It's super easy and does not require creating an issue - just PR with documentation update is enough (and you can create the PR automatically by clicking "Suggest a change on this page" at the bottom right. This will open PR and you can become one of the > 2100 contributors to Airlfow easily.

**Reply by Bowrna** [reply]
I am investigating on this too @potiuk 
It's not working as expected at the DAG level. Maybe I have configured things wrongly? But I will check on this part and see if I can update the documentation part
